R/recycle.R

```diff
@@ -1,31 +1,126 @@
+#'
+#' If vectors aren't size 1, they must all be the same size:
+#'
+#' ```
+#' # This is fine, all vectors are the same size
+#' vec_recycle_common(1:3, 4:6)
+#' #> [[1]]
+#' #> [1] 1 2 3
+#' #>
+#' #> [[2]]
+#' #> [1] 4 5 6
 #'
-#' `vec_recycle(x, size)` recycles a single vector to given size.
-#' `vec_recycle_common(...)` recycles multiple vectors to their common size.
-#' All functions obey the vctrs recycling rules, described below, and will
-#' throw an error if recycling is not possible. See [vec_size()] for the
-#' precise definition of size.
+#' # This is an error
+#' vec_recycle_common(1:3, 4:7)
+#' #> Error:
+#' #> ! Can't recycle `..1` (size 3) to match `..2` (size 4).
+#' ```
 #'
-#' @section Recycling rules:
-#' The common size of two vectors defines the recycling rules, and can be
-#' summarise with the following table:
+#' @section Differences with base R:
 #'
-#' \figure{sizes-recycling.png}
+#' The recycling rules described here are stricter than the ones generally used
+#' by base R. With base R, the rules are usually:
 #'
-#' (Note `NULL`s are handled specially; they are treated like empty
-#' arguments and hence don't affect the size)
+#' - If any vector is length 0, the output will be length 0
 #'
-#' This is a stricter set of rules than base R, which will usually
-#' return output of length `max(nx, ny)`, warning if the length of the longer
-#' vector is not an integer multiple of the length of the shorter.
+#' - Otherwise, the output will be length `max(length_x, length_y)`, and a
+#'   warning will be thrown if the length of the longer vector is not an integer
+#'   multiple of the length of the shorter vector.
+#'
+#' ```
+#' # max(2, 4) == 4
+#' # 1:2 is fully recycled to c(1:2, 1:2)
+#' 1:2 + 1:4
+#' #> [1] 2 4 4 6
+#'
+#' # max(3, 4) == 4, with a warning
+#' # 1:3 is partially recycled to c(1:3, 1)
+#' 1:3 + 1:4
+#' #> Warning in 1:3 + 1:4: longer object length is not a multiple of shorter object
+#' #> length
+#' #> [1] 2 4 6 5
+#'
+#' # Length 0 vector overrides any other length
+#' 1 + numeric()
+#' #> numeric(0)
+#' 1:3 + numeric()
+#' #> numeric(0)
+#' ```
+#'
+#' These rules come from the [R Language
+#' Definition](https://cran.r-project.org/doc/manuals/r-release/R-lang.html#Recycling-rules)
+#' and are mainly focused on arithmetic. Outside of arithmetic, these rules are
+#' not followed consistently:
+#'
+#' ```
+#' # Silent recycling
+#' atan2(1:3, 1:2)
+#' #> [1] 0.7853982 0.7853982 1.2490458
+#'
+#' # Taking `max(3, 0)` rather than recycling to length 0
+#' paste(1:3, integer())
+#' #> [1] "1 " "2 " "3 "
+#'
+#' # R 4.0.1 added `recycle0` for this case
+#' paste(1:3, integer(), recycle0 = TRUE)
+#' #> character(0)
+#'
+#' # Erroring rather than recycling
+#' data.frame(1:2, 1:3)
+#' #> Error in base::data.frame(..., stringsAsFactors = stringsAsFactors):
+#' #> arguments imply differing number of rows: 2, 3
+#' ```
+#'
+#' @name vector_recycling_rules
+#' @keywords internal
+NULL
+
+#' Vector recycling
```

Comment:
If you use a tidyverse function in examples above, this could become a vctrs backend section?

Reply:
I'm not quite sure what you mean

Reply:
You'd mention `vec_recycle_common()` here instead of above.

Reply:
Oh so a `@section` that describes how `vec_recycle_common()` is the engine that powers this for r-lib and the tidyverse